docs: guide for porting an html site to gatsby (#18609)
gatsbyjs/gatsby: docs/docs/porting-an-html-site-to-gatsby.md

    <link rel="stylesheet" type="text/css" href="/assets/normalize.css" />
    <link rel="stylesheet" type="text/css" href="/assets/style.css" />
  </head>
  <body>
    <header>
      <a href="/" class="brand-color logo-text">Taylor's Tidy Trees</a>
      <nav>
        <ul>
          <li><a href="/about.html">About</a></li>
          <li><a href="/services/index.html">Services</a></li>
          <li><a href="/who/index.html">Who We Are</a></li>
          <li><a href="/contact.html">Contact</a></li>
        </ul>
      </nav>
    </header>
    <main>
      <h1>Welcome To Taylor's Tidy Trees</h1>
      <h2>We care about trees of all kinds!</h2>
    </main>
  </body>
</html>
```

In the following sections, you'll convert this block of HTML into its equivalent code in Gatsby.

### Assets

In the `/static` folder of your new Gatsby project, you can see a `favicon.ico` file, this is the Gatsby favicon! All files within `/static` will be served at the root of the app, you can put yours in there and watch it show up in the browser. The other image file in this example, `person.png`, should also be moved to the `/static` folder for later use.

> Gatsby negates the need for CSS `<link>` tags for local CSS. Within the `/src/pages/index.js` file, before the page component is defined, you can add a line for importing each CSS file. Gatsby will then efficiently deliver the CSS with your site.

Reviewer: ```suggestion
Gatsby's bundling system negates the need for manual CSS `<link>` tags for local CSS. Within the `/src/pages/index.js` file, before the page component is defined, you can add a line for importing each CSS file. Gatsby will then efficiently deliver the CSS with your site.
```

> Create a folder at `/src/styles` in the Gatsby project. All the css files for the project can be moved into this new folder. Add an import line for each css file to the Gatsby home page:

Reviewer: ```suggestion
Create a folder at `/src/styles` in the Gatsby project. All the CSS files for the project can be moved into this new folder. Add an `import` line for each CSS file to the Gatsby home page:
```

import React from "react"

import "../styles/normalize.css" // highlight-line
import "../styles/style.css" // highlight-line

export default () => <div>Hello world!</div>
```

### Head elements

You might have noticed that the component in `/src/pages/index.js` doesn't include `<html>`, `<head>` or `<body>`. Gatsby makes a default HTML structure for each page and places the output from `/src/pages/index.js` into its body. More `<head>` child elements and HTML attributes are added to the output page with a module called [React Helmet](https://github.com/nfl/react-helmet). React Helmet is added to a Gatsby project in the command line with npm and then to the Gatsby config file:

```shell
npm install --save react-helmet gatsby-plugin-react-helmet
```

Gatsby projects have a config file at `/gatsby-config.js` where site metadata and options can be specified and plugins added. Add a plugin line with `gatsby-plugin-react-helmet` to your config file:

```js:title=/gatsby-site/gatsby-config.js
/**
 * Configure your Gatsby site with this file.
 *
 * See: https://www.gatsbyjs.org/docs/gatsby-config/
 */

module.exports = {
  plugins: ["gatsby-plugin-react-helmet"], // highlight-line
}
```

> Now you can import the `<Helmet>` component to the `index.js` file and place `<header>` & `<main>` elements for the existing HTML. Copy over the contents of the `<head>` tag, You'll no longer need the `<link>` tags for the CSS files. The Gatsby components must have a single root parent in their code structure so one technique is to add a [React Fragment component](https://reactjs.org/docs/fragments.html) around them:

Reviewer: ```suggestion
Now you can import the `<Helmet>` component to the `index.js` file and place `<header>` & `<main>` elements for the existing HTML. Copy over the contents of the `<head>` tag: you'll no longer need the `<link>` tags for the CSS files. The Gatsby components must have a single root parent in their code structure so one technique is to add a [React Fragment component](https://reactjs.org/docs/fragments.html) around them:
```

> ```jsx:title=/src/pages/index.js

Reviewer: Should this be `/gatsby-site/src/pages/index.js`? Same as below.

        Taylor's Tidy Trees
      </Link>
      <nav>
        <ul>
          <li>
            <Link to="/about">About</Link>
          </li>
          <li>
            <Link to="/services">Services</Link>
          </li>
          <li>
            <Link to="/who">Who We Are</Link>
          </li>
          <li>
            <Link to="/contact">Contact</Link>
          </li>
        </ul>
      </nav>
      {/* highlight-end */}
    </header>
    <main>
      <div>Hello world!</div>
    </main>
  </>
)
```

### Page content

The contents of the `<main>` tag can be copied over from `index.html` to your new `index.js` file mostly unchanged. Your pasted HTML needs a small change to be valid: `class` attributes [must be renamed to `className`](https://reactjs.org/docs/dom-elements.html#classname) for usage with React, as `class` is a reserved word in JavaScript.

Opening the site in a browser again at `http://localhost:8000`, you should have a visually complete home page! You'll make the links functional next by porting more pages. Before that, this guide will briefly explore how HTML and JavaScript combine as JSX in a Gatsby application.

> ### HTML and JavaScript

Reviewer: This small word change evokes a bit more detail:

```suggestion
### HTML from JavaScript
```

          </li>
        </ul>
      </nav>
    </header>
    <main>{children}</main>
  </>
)
```

You can now use the `<Layout>` component to create a `/src/who/index.js` page file:

```jsx:title=/gatsby-site/src/pages/who/index.js
import React from "react"
import Layout from "../components/Layout"
import { Link } from "gatsby"

export default () => (
  <Layout breadcrumbs={["Who We Are"]}>
    <h1>Who We Are</h1>
    <h2>These are our staff:</h2>
    <ul>
      <li>
        <Link to="/who/ella-arborist">Ella (Arborist)</Link>
      </li>
      <li>
        <Link to="/who/sam-surgeon">Sam (Tree Surgeon)</Link>
      </li>
      <li>
        <Link to="/who/marin-leafer">Marin (Leafer)</Link>
      </li>
    </ul>
  </Layout>
)
```

The `Who We Are` link in `index.js` should now work! Now use the `<Layout>` component in the `index.js` page file too:

```jsx:title=/gatsby-site/src/pages/index.js
import React from "react"
import Layout from "../components/Layout" // highlight-line

export default () => (
  {/* highlight-start */}
  <Layout>
    <h1>Welcome To Taylor's Tidy Trees</h1>
    <h2>We care about trees of all kinds!</h2>
  </Layout>
  {/* highlight-end */}
)
```

> Have a check that the `Who We Are` link is still working.

Reviewer: If it isn't, do you have any advice for what they should do next to investigate?

Author: Good point, I think the only way it'd go wrong at that point is if elements the Layout is replacing haven't been completely removed:
```suggestion
Have a check that the `Who We Are` link is still working. If not, check all the elements being replaced by the `<Layout>` component has been removed.
```

That might be clear enough with the example above though, maybe the line could just be deleted. What do you think?

Author: Decided to updated with a note to check with the code snippet directly above the line:

>Have a check that the `Who We Are` link is still working. If not, check that the content is wrapped correctly with the `<Layout>` component as shown above.

### Porting other pages

Now it's time for the work to really pay off! Ella's page is a matter of using your `<Layout>` component again and copying in the main content. Don't forget to change `class` to `className`! A `breadcrumbs` prop can be passed to `<Layout>` to set the dynamic page title. Passing props is similar to an attribute on an HTML element:

```jsx:title=/gatsby-site/src/pages/ella-arborist.js
import React from "react"
import Layout from "../components/Layout"
import { Link } from "gatsby"

export default () => (
  {/* highlight-start */}
  <Layout breadcrumbs={["Who We Are", "Ella"}}>
  {/* highlight-end */}
    <h1>Ella - Arborist</h1>
    <h2>Ella is an excellent Arborist. We guarantee it.</h2>
    <div className="bio-card">
      <img
        alt="Comically crude stick person sketch"
        src="/person.png"
      />
      <p>Ella</p>
    </div>
  </Layout>
)
```

The other 2 `Who We Are` pages for Marin and Sam can now be made with a similar structure. Maybe you are even thinking about another component for the Bio Card!

Once `/services` and the root level HTML files are ported, here is what the finished Gatsby project file structure looks like:

```

> gatsby-site

Reviewer: Love this! I think it will be very helpful.

├── static
│   ├── favicon.ico
│   └── person.png
├── src
│   ├── styles
│   │   ├── normalize.css
│   │   └── style.css
│   ├── components
│   │   └── Layout.js
│   └── pages
│       ├── index.js
│       ├── about.js
│       ├── contact.js
│       ├── 404.js
│       ├── who
│       │   ├── index.js
│       │   ├── ellla-arborist.js
│       │   ├── marin-leafer.js
│       │   └── sam-surgeon.js
│       └── services
│           ├── index.js
│           ├── growing.js
│           ├── cleaning.js
│           └── shrinking.js
├── node_modules
├── .gitignore
├── .prettierignore
├── .prettierrc
├── gatsby-config.js
├── LICENSE
├── package-lock.json
├── package.json
└── README.md
```

## Building & Deploying

With your new Gatsby application taking shape, it's time to integrate it into your existing HTML website.

### Gatsby build step

> You now have a site that mirrors the existing HTML site section. Stop the development server if it's still running; it's time to run the production build! 🎉

Reviewer: ```suggestion
With all of the pages ported over, you now have a site that mirrors the existing HTML site. Stop the development server if it's still running; it's time to run the production build! 🎉
```

In case they missed the note about porting over `/services`, this gives them another nudge to catch up. What do you think?

```shell
gatsby build
```

Once a build is complete, the compiled set of files can be found in `/public`.

### Hosting the new website

> Once built, the contents of the `/public` folder are ready to be hosted at the root (`/`) of a domain. The files can be deployed in exactly the same way that your existing HTML site was deployed. For more deployment options, see the [Deploying and Hosting page](/docs/deploying-and-hosting/).

Reviewer: I don't think we'll know exactly how their site was deployed, and dragging the public folder into FTP can be a bit of a pain. Perhaps we could make it more open-ended?
```suggestion
Once built, the contents of the `/public` folder are ready to be hosted at the root (`/`) of a domain. The files can be deployed in similar ways to how your existing HTML site may have been deployed. For more deployment options including automation with git and cloud services, see the [Deploying and Hosting page](/docs/deploying-and-hosting/).
```

If the Gatsby site is to be hosted at a non-root path, e.g. `example.com/blog/`, Gatsby needs to be informed so page and asset links in the built output can be prefixed.

> [Path prefix](/docs/path-prefix/) is the Gatsby option for setting a non-root hosting path. Here is the example project `gatsby-config.js` file with a path prefix added for `/blog`:

Reviewer: This is nice! 👍

```js:title=/gatsby-site/gatsby-config.js
/**
 * Configure your Gatsby site with this file.
 *
 * See: https://www.gatsbyjs.org/docs/gatsby-config/
 */

module.exports = {
  pathPrefix: "/blog", // highlight-line
  plugins: ["gatsby-plugin-react-helmet"],
}
```

If the HTML and non-HTML files of the Gatsby build output are to be served in different locations, this can also be supported with the [Asset Prefix](/docs/asset-prefix/) option.

Once prefix options have been set, the build must be run with the `--prefix-paths` flag to apply them:

```shell
gatsby build --prefix-paths
```

> 

Reviewer: What do you think of including a tip about gatsby-serve? I could see this tripping people up otherwise.
````suggestion

You can also serve the built site locally for review using the `prefix-paths` flag:

```shell
gatsby serve --prefix-paths
```
````

Author: Excellent! I didn't know about serve 👍

### New website file structure

Here is the structure of the HTML & non-JavaScript asset files in the Gatsby build output:

```
website-domain
  ├── favicon.ico
  ├── person.png
  ├── index.html
  ├── 404
  │   └── index.html
  ├── about
  │   └── index.html
  ├── contact
  │   └── index.html
  ├── services
  │   ├── index.html
  │   ├── growing
  │   │   └── index.html
  │   ├── cleaning
  │   │   └── index.html
  │   ├── shrinking
  │   │   └── index.html
  └── who
      ├── index.html
      ├── ellla-arborist
      │   └── index.html
      ├── marin-leafer
      │   └── index.html
      └── sam-surgeon
          └── index.html
```

## Next steps

> Gatsby can handle images through direct imports to page and component files too! The [asset import documentation](/docs/importing-assets-into-files/) covers this. Once assets are handled through Gatsby, plugins can be used to optimize their processing and delivery.

Reviewer: ```suggestion
Gatsby can handle images through direct imports to page and component files too! The [asset import documentation](/docs/importing-assets-into-files/) covers this. There is also the [Gatsby Image](/docs/gatsby-image/) component for even deeper optimizations. Once assets are handled through Gatsby, plugins can be used to optimize their processing and delivery.
```
I feel like we should mention Gatsby Image here too...

The [building with components doc](/docs/building-with-components/) has information about why Gatsby uses React component architecture and how it fits into a Gatsby application.

> [Sourcing content and data](/docs/content-and-data/) is a great next step if you are interested in separating your content from your website code, such as sourcing the site title from `gatsby-config.js` with GraphQL.

Reviewer: ```suggestion
[Sourcing content and data](/docs/content-and-data/) is a great next step if you are interested in separating your content from your website code, such as sourcing the site title from `gatsby-config.js` with GraphQL and writing content in Markdown.
```
